MicroProfile40 blog to staging (#1218)
OpenLiberty/blogs: posts/2021-03-24-microprofile40-open-liberty-21003.adoc

---
layout: post

> title: "MicroProfile 4.0 support in Open Liberty 21.0.0.3"

Reviewer: This title is too similar to GA release post and I'd like to be a little more descriptive about the fact that this blog is taking a closer look at the APIs.

Possible alternative suggestions:
"A Deep Dive into MicroProfile 4.0 with Open Liberty"
"Making the most of MicroProfile 4.0 with Open Liberty"
"Taking a deeper look at what's new in MicroProfile 4.0 with Open Liberty"
"What's new in MicroProfile 4.0 with Open Liberty"

> title: "MicroProfile 4.0 support in Open Liberty 21.0.0.3"

Reviewer 2: ```suggestion
title: "A Deep Dive into MicroProfile 4.0 with Open Liberty"
```

categories: blog
author_picture: https://avatars.githubusercontent.com/u/10999780?s=460&u=1a66e468c70e84f54bf5a75c27881cc6133ccc20&v=4
author_github: https://github.com/Emily-Jiang

> seo-title: MicroProfile 4.0 support in Open Liberty 21.0.0.3 - OpenLiberty.io

Reviewer: please change this to reflect the title change above

> seo-title: MicroProfile 4.0 support in Open Liberty 21.0.0.3 - OpenLiberty.io

Reviewer 2: ```suggestion
seo-title: A Deep Dive into MicroProfile 4.0 with Open Liberty
```

> seo-description: This post demonstrates how to use MicroProfile 4.0 in Open Liberty 21.0.0.3.

Reviewer: The latest release of MicroProfile 4.0 is the first release under the newly-formed MicroProfile Working Group, delivering alignment with Jakarta EE 8 as well as new and exciting updates to many of the core MicroProfile APIs: Config, Fault Tolerance, Health, JWT RBAC, Metrics, Open API, Open Tracing, Rest Client. This post dives into these updates in further detail and demonstrates how you can utilize MicroProfile 4.0 in Open Liberty 21.0.0.3.

I have extended and updated this description to hopefully increase our SEO.

blog_description: "This post demonstrates how to use MicroProfile 4.0 in Open Liberty 21.0.0.3."
---

> = MicroProfile 4.0 support in Open Liberty 21.0.0.3

Reviewer: From Line 11 it should be:
```suggestion
= MicroProfile 4.0 support in Open Liberty 21.0.0.3
Emily Jiang <https://github.com/Emily-Jiang>
:imagesdir: /
:url-prefix:
:url-about: /
```

Emily Jiang <https://github.com/Emily-Jiang>

> The latest release of link:https://github.com/eclipse/microprofile/releases/tag/4.0.1[MicroProfile 4.0] aligns with link:https://jakarta.ee/release/8/[Jakarta EE 8]. There are many new and exciting features introduced in MicroProfile 4.0. In this post, we'll

Reviewer: Maybe extend and change to this to give more background/intro to MP 4.0.

The latest release of MicroProfile 4.0 is the first release under the newly-formed MicroProfile Working Group, delivering alignment with Jakarta EE 8 (add link) as well as new and exciting updates to many of the core MicroProfile APIs: Config, Fault Tolerance, Health, JWT RBAC, Metrics, Open API, Open Tracing, Rest Client. In this post, we'll dive into these updated features in further detail and demonstrates how you can utilize them in Open Liberty 21.0.0.3.

take a look at all of the MicroProfile 4.0 updated features and then show you how to use them in Open Liberty.

> 

Reviewer: Could be nice here to perhaps add a section about why MicroProfile was delayed and then intro the working group and the fact that IBM is a part of that, helping to continue MP's development, etc, oss, etc...

== What's New In MicroProfile 4.0?

> The full list of MicroProfile 4.0 specification includes the following specification update.

Reviewer: Would maybe be nice to have the short paragraph but below this list here (part about Jakarta EE features being functionally identical to Java EE counterparts. (Note: functional in paragraph below should be functionally. `
``
Would also be nice to have a short sentence here as to why this is important and what it really means? No real intro as to why Jakarta EE 8 is being mentioned here at the moment.

> The full list of MicroProfile 4.0 specification includes the following specification update.

Reviewer 2: ```suggestion
The full list of MicroProfile 4.0 specification includes the update from Jakarta EE 8 and the MicroProfile specification update.
```

=== Jakarta EE 8 Specification updates:

. link:https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html[Jakarta Contexts and Dependency Injection 2.0]
. link:https://jakarta.ee/specifications/annotations/1.3/annotations-spec-1.3.html[Jakarta Annotations 1.3]
. link:https://jakarta.ee/specifications/restful-ws/2.1/restful-ws-spec-2.1.html[Jakarta RESTful Web Services 2.1]
. link:https://jakarta.ee/specifications/jsonb/1.0/jsonb-spec-1.0.html[Jakarta JSON Binding 1.0]
. link:https://jakarta.ee/specifications/jsonp/1.1/jsonp-spec-1.1.html[Jakarta JSON Processing 1.1]

> Open Liberty releases (19.0.0.6 and onwards) are link:https://jakarta.ee/compatibility/#tab-8[Jakarta EE 8 compatible]. Jakarta EE 8 features are functional identical to their Java EE 8 counterparts with the only difference being the maven coordinates.

Reviewer 2: ```suggestion
Open Liberty releases (19.0.0.6 and onwards) are link:https://jakarta.ee/compatibility/#tab-8[Jakarta EE 8 compatible]. Jakarta EE 8 features are functionally identical to their Java EE 8 counterparts with the only difference being the maven coordinates. The alignment with Jakarta EE 8 demonstrates MicroProfile welcomes the new releases from Jakarta EE and tries to keep up with the new releases from Jakarta EE.
```

=== MicroProfile Component Specification Updates:

. link:https://github.com/eclipse/microprofile-config/releases/tag/2.0[MicroProfile Config 2.0]

. link:https://github.com/eclipse/microprofile-fault-tolerance/releases/tag/3.0[MicroProfile Fault Tolerance 3.0]
. link:https://github.com/eclipse/microprofile-health/releases/tag/3.0[MicroProfile Health 3.0]
. link:https://github.com/eclipse/microprofile-jwt-auth/releases/tag/1.2[MicroProfile JWT Authentication 1.2]
. link:https://github.com/eclipse/microprofile-metrics/releases/tag/3.0[MicroProfile Metrics 3.0]
. link:https://github.com/eclipse/microprofile-open-api/releases/tag/2.0[MicroProfile OpenAPI 2.0]
. link:https://github.com/eclipse/microprofile-opentracing/releases/tag/2.0[MicroProfile OpenTracing 2.0]
. link:https://github.com/eclipse/microprofile-rest-client/releases/tag/2.0[MicroProfile Rest Client 2.0]

In the next section, we will go through each MicroProfile Specification and take a further look at their new and exciting features.

== Getting started

> MicroProfile 4.0 implementation is new in Open Liberty 21.0.0.3 - you can use any version from 21.0.0.3 and onwards.

Reviewer: Please add link to GA release blog and/or release page

Maybe we could change this to be "MicroProfile 4.0 implementation is now available in Open Liberty 21.0.0.3....."?

Reviewer 2: What's the link for the GA release page?

> MicroProfile 4.0 implementation is new in Open Liberty 21.0.0.3 - you can use any version from 21.0.0.3 and onwards.

Reviewer 2: ```suggestion
MicroProfile 4.0 implementation is now available in Open Liberty 21.0.0.3 and future releases.
```

> First, you need to add the following link:{url-prefix}/guides/maven-intro.html[Maven] dependency

Reviewer: Have edited this slightly to match the GA release post and also make it clear that both steps are needed:

To make use of MicroProfile 4.0 you must add the feature to both pom.xml and server.xml.
Add the following link:{url-prefix}/guides/maven-intro.html[Maven] dependency to your pom.xml:

Reviewer 2: great thanks!

to your pom.xml:

[source,xml]
----

<dependency>
    <groupId>org.eclipse.microprofile</groupId>
    <artifactId>microprofile</artifactId>
    <version>4.0.1</version>
    <type>pom</type>
    <scope>provided</scope>
</dependency>

----

or, if you use link:{url-prefix}/guides/gradle-intro.html[Gradle]:
[source,gradle]
----
dependencies {
    mpRestClient group: 'org.eclipse.microprofile', name: 'microprofile', version: '4.0.1'
}
----

> Please note: you need to use the version `4.0.1` as the version `4.0` does not pull in any dependencies by mistake (see link:https://github.com/eclipse/microprofile/issues/205[here] for more details).

Reviewer: Not sure I understand what "by mistake" is referring to here? Was this an error in the MP 4.0 release? Perhaps we could make this clearer or even leave out the mention of the mistake and let them visit the link provided for more details?

> Please note: you need to use the version `4.0.1` as the version `4.0` does not pull in any dependencies by mistake (see link:https://github.com/eclipse/microprofile/issues/205[here] for more details).

Reviewer 2: ```suggestion
Please note: you need to use the version `4.0.1` as the version `4.0` does not pull in any dependencies (see link:https://github.com/eclipse/microprofile/issues/205[here] for more details).
```

Also, make sure to configure your Liberty server with the `microProfile-4.0` feature in the server.xml:
[source,xml]
----
<server>
  <featureManager>
    <feature>microProfile-4.0</feature>
    <!-- ... -->
  </featureManager>
  <!-- ... -->
</server>
----

> That's it! Now that we've got our development and deployment environments set up, it's time to dive into each individual MicroProfile Specification to learn the amazing updates!

Reviewer: Can we change this part to...
it's time to dive into each individual MicroProfile Specification to learn more about these highly anticipated updates!

> That's it! Now that we've got our development and deployment environments set up, it's time to dive into each individual MicroProfile Specification to learn the amazing updates!

Reviewer 2: ```suggestion
It's time to dive into each individual MicroProfile Specification to learn more about these highly anticipated updates!
```

. <<config-2.0, What's new in MicroProfile Config 2.0>>
. <<fault-tolerance-3.0, New features in MicroProfile Fault Tolerance 3.0>>
. <<health-3.0, New changes in MicroProfile Health 3.0>>
. <<metrics-3.0, What's New in MicroProfile Metrics 3.0>>
. <<open-tracing-2.0, New updates in  MicroProfile Open Tracing 2.0>>
. <<open-api-2.0, What's New in MicroProfile Open API 2.0>>
. <<rest-client-2.0, Awesome features in MicroProfile Rest Client 2.0>>
. <<jwt-1.2, Must known features in MicroProfile JWT Auth 1.2>>

[#config-2.0]
== What's New in MicroProfile Config 2.0

Quite a few nice capabilities were provided by MicroProfile Config 2.0, such as Config Profile, Config Properties, ConfigValue support, and more.

Open Liberty's MicroProfile Config 2.0 feature is based on link:https://github.com/smallrye/smallrye-config[SmallRye's implementation] of the specification.

Watch this space for a new blog post dedicated to MicroProfile Config 2.0.

[#fault-tolerance-3.0]
== What's New in MicroProfile Fault Tolerance 3.0

> MicroProfile Fault Tolerance allows you to easily apply strategies for mitigating failure to their code. It provides annotations which you can add to methods to use bulkhead, circuit breaker, retry, timeout and fallback strategies.

Reviewer: First sentence references you and then their, this is confusing. Is it for mitigating failures to your code? Should their actually be your?

> MicroProfile Fault Tolerance allows you to easily apply strategies for mitigating failure to their code. It provides annotations which you can add to methods to use bulkhead, circuit breaker, retry, timeout and fallback strategies.

Reviewer 2: ```suggestion
MicroProfile Fault Tolerance allows you to easily apply strategies for mitigating failure to your code. It provides annotations which you can add to methods to use bulkhead, circuit breaker, retry, timeout and fallback strategies.
```

Reviewer 3: suggested edit:

It provides annotations that you can add to methods to use bulkhead, circuit breaker, retry, timeout and fallback strategies.

Fault Tolerance 3.0 introduces two main features.

=== MicroProfile Metrics tags

MicroProfile Fault Tolerance 3.0 overhauls the metrics that are automatically exported, to take advantage of tags. Tags make it easier to use those metrics, to understand when and where your application is failing.

Information which was previously included in the metric name is now included as metric tags. This makes it much easier to query for data from multiple methods and pick out those which may be causing an issue.

Example:

Old metric: `application:ft.<name>.timeout.callsTimedOut.total`

> New metric: `base:ft.timeout.calls.total{method="<name>", timedOut="true"}``

Reviewer: Double backtick ("``") at the end, is this correct?

Reviewer 2: well spotted

Reviewer 2: ```suggestion
New metric: `base:ft.timeout.calls.total{method="<name>", timedOut="true"}`
```

=== Clarify the scope of Bulkheads and CircuitBreakers

> The specification now requires that circuit breakers and bulkheads are singletons, identified by the bean class and the guarded method. For example, if a @RequestScoped bean has a @CircuitBreaker method, all invocations of that method will share the same circuit breaker state, even though each request has a different instance of the bean.

Reviewer: Should we not make "@RequestScoped" and "@CircuitBreakerr" inline code blocks? e.g: @RequestScoped and @CircuitBreaker @tevans78  @GraceJansen 

```suggestion
The specification now requires that circuit breakers and bulkheads are singletons, identified by the bean class and the guarded method. For example, if a `@RequestScoped` bean has a `@CircuitBreaker` method, all invocations of that method will share the same circuit breaker state, even though each request has a different instance of the bean.
```

Get an introduction to MicroProfile Fault Tolerance with the Open Liberty guides link:{url-prefix}/guides/retry-timeout.html[Failing fast and recovering from errors] and link:{url-prefix}/guides/circuit-breaker.html[Preventing repeated failed calls to microservices].

==== Try it now

Enable Fault Tolerance 3.0 and CDI in the `server.xml`, along with any other features you're using.

[source, xml]
----
<featureManager>
  <feature>mpFaultTolerance-3.0</feature>
  <feature>cdi-2.0</feature>
  <feature>jaxrs-2.1</feature>
</featureManager>
----

For more information:

* link:https://github.com/eclipse/microprofile-fault-tolerance/releases/tag/3.0[MicroProfile Fault Tolerance 3.0 Release Page (Javadoc & Spec)]

* link:https://download.eclipse.org/microprofile/microprofile-fault-tolerance-3.0/microprofile-fault-tolerance-spec-3.0.html#release_notes_30[Release Notes]

[#metrics-3.0]

> === What's New in MicroProfile Metrics 3.0

Reviewer: This text should be larger - should be two equals signs not three

> === What's New in MicroProfile Metrics 3.0

Reviewer 2: ```suggestion
== What's New in MicroProfile Metrics 3.0
```

MicroProfile Metrics 3.0 introduces new metric values for the existing SimpleTimer and Timer metrics. Additionally a new REST metric is introduced for better monitoring and handling of unmapped exceptions. A notable change to the MicroProfile Metrics programming model regarding CDI Producers has been made. Lastly a medley of API improvements and refactoring have been added in this release.

> ==== SimpleTimer

Reviewer: Should these subtitles be the same size as the subtitles in the previous section on Fault Tolerance? At the moment they are smaller

> ==== SimpleTimer

Reviewer 2: ```suggestion
=== SimpleTimer
```

The SimpleTimer metric now tracks and reports the highest and lowest recorded time duration of the previous complete minute. See link:https://download.eclipse.org/microprofile/microprofile-metrics-3.0/apidocs/index.html?org/eclipse/microprofile/metrics/SimpleTimer.html[SimpleTimer Javadoc] for more information.

==== Timer

> The Time metric now tracks and reports the total elapsed time duration. See link:https://download.eclipse.org/microprofile/microprofile-metrics-3.0/apidocs/index.html?org/eclipse/microprofile/metrics/Timer.html[Timer Javadoc] for more information.

Reviewer: Should this say "The Time metric" or "The Timer metric"?

> The Time metric now tracks and reports the total elapsed time duration. See link:https://download.eclipse.org/microprofile/microprofile-metrics-3.0/apidocs/index.html?org/eclipse/microprofile/metrics/Timer.html[Timer Javadoc] for more information.

Reviewer 2: ```suggestion
The Timer metric now tracks and reports the total elapsed time duration. See link:https://download.eclipse.org/microprofile/microprofile-metrics-3.0/apidocs/index.html?org/eclipse/microprofile/metrics/Timer.html[Timer Javadoc] for more information.
```

> ==== REST metric

Reviewer 2: ```suggestion
=== REST metric
```

> A new `REST.request.unmappedException.total` metric that is backed by a counter metric has been introduced. Similar to the `REST.request` metric, there will be one unique metric for each REST endpoint identified by a class and method label. The new REST metric will count the amount of times the request ends in an unmapped exception. The REST.request metric corresponding to this REST endpoint will not record any values if an unmapped exception has occured.

Reviewer: ```suggestion
A new `REST.request.unmappedException.total` metric that is backed by a counter metric has been introduced. Similar to the `REST.request` metric, there will be one unique metric for each REST endpoint identified by a class and method label. The new REST metric will count the amount of times the request ends in an unmapped exception. The `REST.request` metric corresponding to this REST endpoint will not record any values if an unmapped exception has occured.
```

==== CDI Producer
The `@Metrics` annotation will no longer support the method target (i.e it can not be annotated on a method). Additionally, it will not support usage with CDI Producers.

==== Try it now

Enable Metrics 3.0 in the `server.xml`, along with any other features you're using.

[source, xml]
----
<featureManager>
    <feature>mpMetrics-3.0</feature>
</featureManager>
----

More information:

* link:https://github.com/eclipse/microprofile-metrics/releases/tag/3.0[MicroProfile Metrics 3.0 Release Page (Javadoc & Spec)]

* link:https://download.eclipse.org/microprofile/microprofile-metrics-3.0/microprofile-metrics-spec-3.0.html#release_notes_3_0[Release Notes]

[#health-3.0]

> === New changes in MicroProfile Health 3.0

Reviewer: Again, think this should be larger text - should be two equals signs not three

MicroProfile Health 3.0 enables you to provide your own health check procedures to be invoked by Open Liberty, to verify the health of your microservices.

MicroProfile Health allows services to report their health, and publish overall health status to a defined endpoint. A service reports "UP" if it is available and reports "DOWN" if it is unavailable. In addition to reporting the state of individual services, MicroProfile Health also aggregates the overall status as "UP" if all the services are "UP". If one of the services is "DOWN" then the overall status is also "DOWN". A service orchestrator can then use the health statuses to make decisions.

MicroProfile Health checks its own health by performing necessary self-checks and then reports its overall status by implementing the API provided by MicroProfile Health. A self-check can be a check on anything that the service needs, such as a dependency, a successful connection to an endpoint, a system property, a database connection, or the availability of required resources. MicroProfile offers checks for both liveness and readiness.

Significant changes in MicroProfile Health 3.0:

> * The overall default Readiness status was changed to "DOWN", with an empty response until all the deployed application(s) have been started. A new MicroProfile Config property (`mp.health.default.readiness.empty.response=UP`) is introduced to change the overall default Readiness check status to "UP", during application start up, that do not have any user-defined health checks.

Reviewer: ```suggestion
* The overall default Readiness status was changed to "DOWN", with an empty response until the deployed application(s) have been started. A new MicroProfile Config property (`mp.health.default.readiness.empty.response=UP`) is introduced to change the overall default Readiness check status to "UP", during application start up, that do not have any user-defined health checks.
```

Author: Actually I'd be tempted to put the word _all_ in italics rather than remove it

Reviewer 3: Here's how we have it in the docs:

In MicroProfile Health versions 2.0 and later, the overall default readiness status is `DOWN`, with an empty response, until all the deployed applications in Open Liberty are started. In MicroProfile Health versions 3.0 and later, the `mp.health.default.readiness.empty.response=UP` MicroProfile Config property is available. You can specify this property to change the overall default readiness status to `UP` while deployed applications are starting up.

* The `HealthCheckResponseBuilder.state(Boolean UP)` method was also renamed to `HealthCheckResponseBuilder.status(Boolean UP)` for HealthCheckResponse deserialization compatibility, where the JSON health check response string can now be deserialized into an HealthCheckResponse object.

* The deprecated `@Health` qualifier was removed. Instead, the `@Liveness` or `@Readiness` qualifiers should be used as appropriate in your HealthCheck implementations.

> Applications are expected to provide health check procedures by implementing the HealthCheck interface with the `@Liveness` or `@Readiness` annotations. These are used by Open Liberty to verify the Liveness or Readiness of the application, respectively. Add the logic of your health check in the call() method, and return an instance of `HealthCheckResponse` by using the simple `up` and `down` API methods:

Reviewer: ```suggestion
Applications are expected to provide health check procedures by implementing the HealthCheck interface with the `@Liveness` or `@Readiness` annotations. These are used by Open Liberty to verify the Liveness or Readiness of the application, respectively. Add the logic of your health check in the `call()` method, and return an instance of `HealthCheckResponse` by using the simple `up` and `down` API methods:
```

[source, java]
----
//**Liveness Check**
@Liveness
@ApplicationScoped
public class AppLiveCheck implements HealthCheck {
...
    @Override
     public HealthCheckResponse call() {
       ...
       HealthCheckResponse.up("my-liveness-check");
       ...
     }
}

//**Readiness Check**
@Readiness
@ApplicationScoped
public class AppReadyCheck implements HealthCheck {
...
    @Override
     public HealthCheckResponse call() {
       ...
       HealthCheckResponse.named("my-app-readiness").status(isMyAppReady()).build();
       ...
     }
}
...
----

> To view the status of each health check, access the either the `http://<hostname>:<port>/health/live` or `http://<hostname>:<port>/health/ready endpoints`.

Reviewer: This sentence does not make sense - "access the either the"?

Reviewer 2: ```suggestion
To view the status of each health check, access either `http://<hostname>:<port>/health/live` or `http://<hostname>:<port>/health/ready endpoints`.
```

More information:

* link:https://github.com/eclipse/microprofile-health/releases/tag/3.0[MicroProfile Health 3.0 Release Page (Javadoc & Spec)]

* link:https://download.eclipse.org/microprofile/microprofile-health-3.0/microprofile-health-spec-3.0.html#release_notes_3_0[Release notes]

[#opentracing-2.0]

> === New updates in  MicroProfile Open Tracing 2.0

Reviewer: Again, think this is too small - two equals rather than three

MicroProfile OpenTracing 2.0 can be used to profile and monitor applications built using microservice architecture.

MicroProfile OpenTracing 2.0 has upgraded the OpenTracing API to version 0.33.0.  This allows the
use of tracing backends and their libraries that are built on OpenTracing API 0.33.0.

==== Try it now

Include the following in the `server.xml`:

[source, xml]
----
    <feature>mpOpenTracing-2.0</feature>
----

> Also configure a tracing backend such as Jaeger or Zipkin. +

Reviewer: Should this say "you can also"?

Reviewer 2: ```suggestion
Next is to configure a tracing backend such as Jaeger or Zipkin. +
```

For Jaeger, add the following maven dependencies in the application's pom.xml.

[source, xml]
----
<dependency>
    <groupId>io.jaegertracing</groupId>
    <artifactId>jaeger-client</artifactId>
    <version>1.2.0</version>
</dependency>
<dependency>
    <groupId>org.slf4j</groupId>
    <artifactId>slf4j-api</artifactId>
    <version>1.7.30</version>
</dependency>
<dependency>
    <groupId>org.slf4j</groupId>
    <artifactId>slf4j-jdk14</artifactId>
    <version>1.7.30</version>
</dependency>
----

> You can find out more about about configuring Jaeger settings using environment variables by looking

Reviewer: We also have a guide on enabling distributed tracing using Jaeger [[here](https://openliberty.io/guides/microprofile-opentracing-jaeger.html)]

at link:https://github.com/jaegertracing/jaeger-client-java/blob/v1.2.0/jaeger-core/README.md[jaeger-client-java readme].

For the `JAEGER_PASSWORD` environment variable, the password can be encoded using the `securityUtility` command.

Depending on Jaeger’s sampling settings `JAEGER_SAMPLER_TYPE` and `JAEGER_SAMPLER_PARAM`,
Jaeger may not report every span created by the applications.

> For Zipkin, take a look at the link:https://github.com/WASdev/sample.opentracing.zipkintracer[sample project] to see how to implement a tracer for Liberty.

Reviewer: Also have a guide for Zipkin too ([here](https://openliberty.io/guides/microprofile-opentracing.html))

Define your application in the `server.xml`:
[source, xml]
----
<webApplication location="yourapp.war" contextRoot="/yourapp">
    <!-- enable visibility to third party APIs -->
    <classloader apiTypeVisibility="+third-party" />
</webApplication>
----

Once you have hit some JAX-RS endpoints of your application, you should be able to find spans in the user interface of your tracing backend.

More information:

* link:https://github.com/eclipse/microprofile-opentracing/releases/tag/2.0[MicroProfile OpenTracing 2.0 Release Page (Javadoc & Spec)]

* link:https://download.eclipse.org/microprofile/microprofile-opentracing-2.0/microprofile-opentracing-spec-2.0.html#_release_2_0[Release notes]

[#open-api-2.0]

> === What's New in MicroProfile Open API 2.0

Reviewer: Should this be larger text? two rather than three equals signs?

Reviewer 2: ```suggestion
== What's New in MicroProfile Open API 2.0
```

Shortly we will be publishing a dedicated blog post solely on MicroProfile Rest Client 2.0, which will demonstrate all the new features and how to use them.

[#jwt-1.2]
=== Enhancements in MicroProfile JWT Auth 1.2

MicroProfile JWT 1.2 simplifies the configuration for managing the validation of the JWT by introducing new MicroProfile Config properties. Enhanced signature algorithm support is added in this Open Liberty implementation.

MicroProfile JWT 1.2 adds a number of MicroProfile Config properties to control different aspects of the JWT validation. The specification also adds support for the ES256 signature algorithm. The Open Liberty implementation additionally supports using the RS384, RS512, HS384, HS512, ES256, ES384, and the ES512 signature algorithms.

MicroProfile JWT 1.2 also supports the JWT specified in a cookie.

Watch this space for a more detailed blog on MicroProfile JWT 1.2.

> == Summary

Reviewer: I think we should add a shortcut here and link it to the list aswell since it's a h2 heading

```suggestion
[#summary]
== Summary
```

> MicroProfile 4.0 has many new updates as shown above. It is time to try out MicroProfile 4.0 using Open Liberty 21.0.0.3!

Reviewer: Can we change this to: MicroProfile 4.0 offers many new and exciting updates as shown above.

Can we link to the release page or GA blog here?

Reviewer 2: Great idea!

Reviewer 3: For accessibility compliance, avoid directional language, like "above". An alternative could be:

As you can see, MicroProfile 4.0 has many new updates.

> As always, let us know if you have any questions with this new feature. Thanks for checking it out!

Reviewer: Can we have a link to stack overflow for questions like we do on the GA blog?

Reviewer 2: Good idea!
